lib/server.js

```diff
@@ -41,123 +207,59 @@ function createServer (options, httpHandler) {
+  if (args.length === 0) {
+    return { port: 0, host: 'localhost' }
+  }
 
-    const listenPromise = (listenOptions) => {
-      if (this[kState].listening && this[kState].closing) {
-        return Promise.reject(new FST_ERR_REOPENED_CLOSE_SERVER())
-      } else if (this[kState].listening) {
-        return Promise.reject(new FST_ERR_REOPENED_SERVER())
-      }
+  const cb = typeof args[args.length - 1] === 'function' ? args.pop() : undefined
+  const options = { cb: cb }
 
-      return this.ready().then(() => {
-        let errEventHandler
-        const errEvent = new Promise((resolve, reject) => {
-          errEventHandler = (err) => {
-            this[kState].listening = false
-            reject(err)
-          }
-          server.once('error', errEventHandler)
-        })
-        const listen = new Promise((resolve, reject) => {
-          server.listen(listenOptions, () => {
-            server.removeListener('error', errEventHandler)
-            resolve(logServerAddress())
-          })
-          // we set it afterwards because listen can throw
-          this[kState].listening = true
-        })
+  const firstArg = args[0]
+  const argsLength = args.length
+  const lastArg = args[argsLength - 1]
+  /* Deal with listen (options) || (handle[, backlog]) */
+  if (typeof firstArg === 'object' && firstArg !== null) {
+    options.backlog = argsLength > 1 ? lastArg : undefined
+    Object.assign(options, firstArg)
+  } else if (typeof firstArg === 'string' && isNaN(firstArg)) {
+    /* Deal with listen (pipe[, backlog]) */
+    options.path = firstArg
+    options.backlog = argsLength > 1 ? lastArg : undefined
+  } else {
+    /* Deal with listen ([port[, host[, backlog]]]) */
+    options.port = argsLength >= 1 && firstArg ? firstArg : 0
+    // This will listen to what localhost is.
+    // It can be 127.0.0.1 or ::1, depending on the operating system.
+    // Fixes https://github.com/fastify/fastify/issues/1022.
+    options.host = argsLength >= 2 && args[1] ? args[1] : 'localhost'
+    options.backlog = argsLength >= 3 ? args[2] : undefined
+  }
 
-        return Promise.race([
-          errEvent, // e.g invalid port range error is always emitted before the server listening
-          listen
-        ])
-      })
-    }
+  return options
+}
 
-    const logServerAddress = () => {
-      let address = server.address()
-      const isUnixSocket = typeof address === 'string'
-      /* istanbul ignore next */
-      if (!isUnixSocket) {
-        if (address.address.indexOf(':') === -1) {
-          address = address.address + ':' + address.port
-        } else {
-          address = '[' + address.address + ']:' + address.port
-        }
-      }
-      /* istanbul ignore next */
-      address = (isUnixSocket ? '' : ('http' + (this[kOptions].https ? 's' : '') + '://')) + address
-      this.log.info('Server listening at ' + address)
-      return address
+function logServerAddress (server) {
+  let address = server.address()
```

Comment:
Will `server.address` return both IP addresses?

Reply:
No, the other server interface is not accessible by the user

Reply:
Why? It's their resources being used. We should be returning all addresses + ports being used by "the server".

Reply:
I think binding all the host's interfaces is a utility to speed up the local development when users call `localhost`.
Returning an array would overcomplicate the development code (2 hosts bind) over the code that runs in production (1 single host bind).

I would hide the complexity to the user in this case because I don't see any benefits.

Reply:
The alias `localhost` has nothing to do with "local development" other than all systems define local loopback addresses, including developer systems. We absolutely should be providing access to all addresses we bind to. Even if that means adding `server.address6()` instead of making `server.address()` return everything.

Reply:
Looks good to me. Thanks.